### PR TITLE
Combine UpgradeNudgePaid and UpgradeNudge

### LINKED ISSUE #66327: Marketplace: Mix both UpgradeNudge and UpgradeNudgePaid in one function
As @Rolilink [here](https://github.com/Automattic/wp-calypso/pull/66197#discussion_r937193646), we should mix both `UpgradeNudge` and `UpgradeNudgePaid` in one function.

File affected: [/client/my-sites/plugins/plugins-discovery-page/index.jsx](https://github.com/Automattic/wp-calypso/blob/3b2fd28a03fcf88ce6b7f2779f0b2f8c46f602bc/client/my-sites/plugins/plugins-discovery-page/index.jsx)

### PR DESCRIPTION
**Pull request by DustyReagan**
#### Proposed Changes

* Combine UpgradeNudgePaid and UpgradeNudge. See: https://github.com/Automattic/wp-calypso/issues/66327

#### Testing Instructions

* Use the Calypso live link below
* Go to `/plugins`.
* On a site without plan it should show the upgrade banner for all plugins image
* On a site with Starter Plan it should not show a banner over the premium plugins but a banner over free plugins image
* On a site with Pro or Business plan it should show no banners at all.
* Sandbox public-api
* In `WPCOM_Features` change the value for `self::INSTALL_PURCHASED_PLUGINS` to include `self::WPCOM_PERSONAL_AND_HIGHER_PLANS`.
* Back in Calypso, on a free site it should now show two upgrade banners. One for the Personal plan over premium plugins and one for all plugins over free plugins image

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Fixes #66327

### CODE REVIEW

**Review by lupus2k — APPROVED** [review]
LGTM!

File: client/my-sites/plugins/plugins-discovery-page/index.jsx [hunk 1]
```diff
@@ -24,63 +24,88 @@ import usePlugins from '../use-plugins';
...
-	if ( ! selectedSite?.ID || ! sitePlan || isVip || jetpackNonAtomic ) {
-		return null;
-	}
-	const isLegacyPlan = isBlogger( sitePlan ) || isPersonal( sitePlan ) || isPremium( sitePlan );
-	const checkoutPlan = eligibleForProPlan && ! isLegacyPlan ? 'pro' : 'business';
-	const bannerURL = `/checkout/${ siteSlug }/${ checkoutPlan }`;
-	const plan = findFirstSimilarPlanKey( sitePlan.product_slug, {
-		type: TYPE_BUSINESS,
-	} );
+	const requiredPlans = useSelector( ( state ) =>
```
Comment (Rolilink) [inline client/my-sites/plugins/plugins-discovery-page/index.jsx]:
```
const requiredPlansPurchasedPlugins = useSelector( ( state ) =>
		getPlansForFeature( state, selectedSite?.ID, WPCOM_FEATURES_INSTALL_PURCHASED_PLUGINS )
	);
```

This seems to be repeated above

File: client/my-sites/plugins/plugins-discovery-page/index.jsx [hunk 2]
```diff
@@ -24,63 +24,88 @@ import usePlugins from '../use-plugins';
...
-	if ( ! requiredPlans ) {
+		bannerURL = `/checkout/${ siteSlug }/${ requiredPlan.getPathSlug() }`;
+		feature = WPCOM_FEATURES_INSTALL_PURCHASED_PLUGINS;
+		plan = requiredPlan.getStoreSlug();
+		title = translate( 'Upgrade to the %(planName)s plan to install premium plugins.', {
+			textOnly: true,
+			args: { planName: requiredPlan.getTitle() },
+		} );
+	} else if (
+		( paidPlugins && ! hasInstallPurchasedPlugins && ! lowerPlanAvailable ) ||
```
Comment (Rolilink) [inline client/my-sites/plugins/plugins-discovery-page/index.jsx]: I think `paidPlugins` could be changed by `isNotJetpackNonAtomic` since is more descriptive about the reason why we are choosing different condition

File: client/my-sites/plugins/plugins-discovery-page/index.jsx [hunk 3]
```diff
@@ -24,63 +24,88 @@ import usePlugins from '../use-plugins';
...
+		( ! paidPlugins && ( hasInstallPurchasedPlugins || lowerPlanAvailable ) )
```
Comment (Rolilink) [inline client/my-sites/plugins/plugins-discovery-page/index.jsx]: Thinking out loud, not actionable:

I am not sure why is this needed in the first place, I know is present in the old code but I don't totally understand what is happening there also .. this basically means that in jetpackNonAtomic sites (hosted jetpack sites) if a site has the ability to install purchased plugins ( business or above) or there is a lower plan available (personal or above) it should show the free one?

Comment (Rolilink) [reply client/my-sites/plugins/plugins-discovery-page/index.jsx]: I think adding a follow-up task for researching this is a good idea!

**Comment by Rolilink** [comment]
LGTM! some small changes and is good for me, let's test this one thoroughly on different types of sites tho

File: client/my-sites/plugins/plugins-discovery-page/index.jsx [hunk 4]
```diff
@@ -146,38 +171,17 @@ const PluginsDiscoveryPage = ( props ) => {
...
-						{ ! hasInstallPurchasedPlugins && ! lowerPlanAvailable && (
-							<UpgradeNudge { ...props } />
-						) }
+						<UpgradeNudge { ...props } paidPlugins={ true } />
 					</div>
 					<PaidPluginsSection { ...props } />
 				</>
 			) }
-			{ ( hasInstallPurchasedPlugins || lowerPlanAvailable ) && <UpgradeNudge { ...props } /> }
+			{ <UpgradeNudge { ...props } /> }
```
Comment (lsl) [inline client/my-sites/plugins/plugins-discovery-page/index.jsx]: Can we make this component so it looks like:
```
...
<UpgradeNudge />
<PaidPluginSection />
<FeaturedPluginSection />
...
```

I'm thinking it'll be easier to follow if we move all the `isJetpack ? null : <Component>` display logic into the components themselves, especially in the nudge case where we're doing further should I null checks.

File: client/my-sites/plugins/plugins-discovery-page/index.jsx [hunk 5]
```diff
@@ -146,38 +171,17 @@ const PluginsDiscoveryPage = ( props ) => {
...
-	// Whether to show an upgrade banner specific to premium plugins.
-	const lowerPlanAvailable =
-		! hasInstallPurchasedPlugins &&
-		requiredPlansPurchasedPlugins[ 0 ] !== requiredPlansAllPlugins[ 0 ];
-
 	return (
 		<>
 			{ ! props.jetpackNonAtomic && (
 				<>
 					<div className="plugins-discovery-page__upgrade-banner">
```
Comment (lsl) [inline client/my-sites/plugins/plugins-discovery-page/index.jsx]: We can remove this div, dead code, though I wonder if it being dead is what caused this: https://github.com/Automattic/wp-calypso/issues/66484

File: client/my-sites/plugins/plugins-discovery-page/index.jsx [hunk 6]
```diff
@@ -24,63 +24,88 @@ import usePlugins from '../use-plugins';
...
+		const requiredPlan = getPlan( requiredPlans[ 0 ] );
 
-	if ( ! requiredPlans ) {
+		bannerURL = `/checkout/${ siteSlug }/${ requiredPlan.getPathSlug() }`;
+		feature = WPCOM_FEATURES_INSTALL_PURCHASED_PLUGINS;
+		plan = requiredPlan.getStoreSlug();
+		title = translate( 'Upgrade to the %(planName)s plan to install premium plugins.', {
+			textOnly: true,
+			args: { planName: requiredPlan.getTitle() },
+		} );
```
Comment (lsl) [inline client/my-sites/plugins/plugins-discovery-page/index.jsx:81]: If the props are just being set once it's usually better to return the component instead of deferring. e.g return a `<UpsellNudge ... />` here instead.

File: client/my-sites/plugins/plugins-discovery-page/index.jsx [hunk 7]
```diff
@@ -24,63 +24,88 @@ import usePlugins from '../use-plugins';
...
+			return null;
+		}
+		const isLegacyPlan = isBlogger( sitePlan ) || isPersonal( sitePlan ) || isPremium( sitePlan );
+		const checkoutPlan = eligibleForProPlan && ! isLegacyPlan ? 'pro' : 'business';
+		bannerURL = `/checkout/${ siteSlug }/${ checkoutPlan }`;
+		feature = FEATURE_INSTALL_PLUGINS;
+		plan = findFirstSimilarPlanKey( sitePlan.product_slug, {
+			type: TYPE_BUSINESS,
+		} );
+		title =
```
Comment (lsl) [inline client/my-sites/plugins/plugins-discovery-page/index.jsx]: I think this might be easier to understand if we move the pro v business question up to the top level if branching and just have three `return <NudgeUpsell .. />`'s.

File: client/my-sites/plugins/plugins-discovery-page/index.jsx [hunk 8]
```diff
@@ -24,63 +24,88 @@ import usePlugins from '../use-plugins';
...
+		title = translate( 'Upgrade to the %(planName)s plan to install premium plugins.', {
+			textOnly: true,
+			args: { planName: requiredPlan.getTitle() },
+		} );
+	} else if (
+		( paidPlugins && ! hasInstallPurchasedPlugins && ! lowerPlanAvailable ) ||
+		( ! paidPlugins && ( hasInstallPurchasedPlugins || lowerPlanAvailable ) )
+	) {
+		if ( ! selectedSite?.ID || ! sitePlan || isVip || jetpackNonAtomic ) {
+			return null;
```
Comment (lsl) [inline client/my-sites/plugins/plugins-discovery-page/index.jsx]: Can we move this to the top of the function and remove the jetpackNonAtomic check the caller is doing?

**Review by lsl — APPROVED** [review]
This combined nudge is looking pretty dense after merging everything into one, dropped a few suggestions on how to maybe help the readability but otherwise tested and LGTM.

Tested as many plan types as I could:
- Jetpack
- Starter
- Pro
- Business
- Premium
- Free

The only one I couldn't figure out if it was working correctly was starter where I didn't see any banner, but my test site did have an existing plugin so maybe working as intended. Removing the subscription/plugin didn't show the banner though - might just be misunderstanding when this should run.

**Review by Rolilink — APPROVED** [review]
LGTM too

File: client/my-sites/plugins/plugins-discovery-page/index.jsx (hunk 4 above)
Comment (DustyReagan) [reply client/my-sites/plugins/plugins-discovery-page/index.jsx]: ✅

File: client/my-sites/plugins/plugins-discovery-page/index.jsx (hunk 5 above)
Comment (DustyReagan) [reply client/my-sites/plugins/plugins-discovery-page/index.jsx]: ✅  - Didn't seem to fix the issue in #66484 though. :/

File: client/my-sites/plugins/plugins-discovery-page/index.jsx (hunk 6 above)
Comment (DustyReagan) [reply client/my-sites/plugins/plugins-discovery-page/index.jsx:81]: ✅

File: client/my-sites/plugins/plugins-discovery-page/index.jsx (hunk 7 above)
Comment (DustyReagan) [reply client/my-sites/plugins/plugins-discovery-page/index.jsx]: ✅

File: client/my-sites/plugins/plugins-discovery-page/index.jsx (hunk 8 above)
Comment (DustyReagan) [reply client/my-sites/plugins/plugins-discovery-page/index.jsx]: ✅

File: client/my-sites/plugins/plugins-discovery-page/index.jsx (hunk 1 above)
Comment (DustyReagan) [reply client/my-sites/plugins/plugins-discovery-page/index.jsx]: ✅ - Removed the duplicate.

File: client/my-sites/plugins/plugins-discovery-page/index.jsx (hunk 2 above)
Comment (DustyReagan) [reply client/my-sites/plugins/plugins-discovery-page/index.jsx]: I was using `paidPlugins` to signify that a particular banner is being used to upsell paid plugins specifically vs free plugins. I'm hoping the most recent commit maybe made the conditional logic easier to follow?

File: client/my-sites/plugins/plugins-discovery-page/index.jsx (hunk 3 above)
Comment (DustyReagan) [reply client/my-sites/plugins/plugins-discovery-page/index.jsx]: The logic is hard to follow, but basically, there are potentially 2 distinct upsells:

1. one for unlocking the feature to install "paid plugins" (the top banner)
2. one for the feature to install "free plugins" (the 2nd banner)

- If it's a Jetpack site, we should show NO banners.
- If the banners would be duplicates, we should only show the top banner, because the upsell applies to both free and paid plugins.
- If the banners to upsell free and paid plugins are different, we should show both.
- If we have the ability to install paid plugins, but not free plugins (Starter plan), we should show the banner upselling free plugin installs (the 2nd banner), but not the top banner.
- If you can install free plugins, you can install paid plugins, so that banner scenario doesn't exist.

**Review by lupus2k — APPROVED** [review]

**Review by Rolilink — APPROVED** [review]

**Comment by Rolilink** [comment]
I wasn't able to test it but overall the changes LGTM!

File: client/my-sites/plugins/plugins-discovery-page/index.jsx [hunk 9]
```diff
@@ -146,38 +191,11 @@ const PluginsDiscoveryPage = ( props ) => {
...
-							<UpgradeNudge { ...props } />
-						) }
-					</div>
-					<PaidPluginsSection { ...props } />
-				</>
-			) }
-			{ ( hasInstallPurchasedPlugins || lowerPlanAvailable ) && <UpgradeNudge { ...props } /> }
+			<UpgradeNudge { ...props } paidPlugins={ true } />
+			<PaidPluginsSection { ...props } />
+			<UpgradeNudge { ...props } />
```
Comment (lsl) [inline client/my-sites/plugins/plugins-discovery-page/index.jsx:198]: Why do we need two of these?

File: client/my-sites/plugins/plugins-discovery-page/index.jsx (hunk 3 above)
Comment (lsl) [reply client/my-sites/plugins/plugins-discovery-page/index.jsx]: Thanks for laying this out, I think this is the condition I was missing:

- If we have the ability to install paid plugins, but not free plugins (Starter plan), we should show the banner upselling free plugin installs (the 2nd banner), but not the top banner.

File: client/my-sites/plugins/plugins-discovery-page/index.jsx (hunk 2 above)
Comment (Rolilink) [reply client/my-sites/plugins/plugins-discovery-page/index.jsx]: it did

**Review by lsl — APPROVED** [review]
This is way easier to follow now, thanks!

File: client/my-sites/plugins/plugins-discovery-page/index.jsx (hunk 3 above)
Comment (Rolilink) [reply client/my-sites/plugins/plugins-discovery-page/index.jsx]: > If the banners would be duplicates, we should only show the top banner, because the upsell applies to both free and paid plugins.

This was what I was missing
